$Step-1$ import the packages

In [1]:
import pandas as pd 
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

$Step-2$ Read the data

In [2]:
df=pd.read_csv("C:\\Users\\lenovo\Desktop\\Data Science course\\Machine learning\\Classification ML with Python(Churn data)\\Preprocessed_data.csv")

In [3]:
df

Gender  Age  Married  Number of Dependents   Latitude   Longitude  \
0          0   37        1                     0  34.827662 -118.999073   
1          1   46        0                     0  34.162515 -118.203869   
2          1   50        0                     0  33.645672 -117.922613   
3          1   78        1                     0  38.014457 -122.115432   
4          0   75        1                     0  34.227846 -119.079903   
...      ...  ...      ...                   ...        ...         ...   
4830       0   53        0                     0  36.807595 -118.901544   
4831       0   20        0                     0  32.759327 -116.997260   
4832       1   40        1                     0  37.734971 -120.954271   
4833       1   22        0                     0  39.108252 -123.645121   
4834       1   21        1                     0  33.001813 -117.263628   

      Number of Referrals  Tenure in Months  Offer  \
0                       2                 9      0   
1                       0                 9      0   
2                       0                 4      5   
3                       1                13      4   
4                       3                 3      0   
...                   ...               ...    ...   
4830                    0                 1      5   
4831                    0                13      4   
4832                    1                22      4   
4833                    0                 2      5   
4834                    5                67      1   

      Avg Monthly Long Distance Charges  ...  Contract  Paperless Billing  \
0                                 42.39  ...         1                  1   
1                                 10.69  ...         0                  0   
2                                 33.65  ...         0                  1   
3                                 27.82  ...         0                  1   
4                                  7.38  ...         0                  1   
...                                 ...  ...       ...                ...   
4830                              42.09  ...         0                  1   
4831                              46.68  ...         1                  0   
4832                              16.20  ...         0                  1   
4833                              18.62  ...         0                  1   
4834                               2.12  ...         2                  0   

      Payment Method  Monthly Charge  Total Charges  Total Refunds  \
0                  1           65.60         593.30           0.00   
1                  1           -4.00         542.40          38.33   
2                  0           73.90         280.85           0.00   
3                  0           98.00        1237.85           0.00   
4                  1           83.90         267.40           0.00   
...              ...             ...            ...            ...   
4830               1           70.15          70.15           0.00   
4831               1           55.15         742.90           0.00   
4832               0           85.10        1873.70           0.00   
4833               1           50.30          92.75           0.00   
4834               1           67.85        4627.65           0.00   

      Total Extra Data Charges  Total Long Distance Charges  Total Revenue  \
0                            0                       381.51         974.81   
1                           10                        96.21         610.28   
2                            0                       134.60         415.45   
3                            0                       361.66        1599.51   
4                            0                        22.14         289.54   
...                        ...                          ...            ...   
4830                         0                        42.09         112.24   
4831                         0                       606.84       

$Step-3$ Divide data input and output

In [4]:
X=df.drop('Customer Status',axis=1)
y=df['Customer Status']

In [5]:
X

Gender  Age  Married  Number of Dependents   Latitude   Longitude  \
0          0   37        1                     0  34.827662 -118.999073   
1          1   46        0                     0  34.162515 -118.203869   
2          1   50        0                     0  33.645672 -117.922613   
3          1   78        1                     0  38.014457 -122.115432   
4          0   75        1                     0  34.227846 -119.079903   
...      ...  ...      ...                   ...        ...         ...   
4830       0   53        0                     0  36.807595 -118.901544   
4831       0   20        0                     0  32.759327 -116.997260   
4832       1   40        1                     0  37.734971 -120.954271   
4833       1   22        0                     0  39.108252 -123.645121   
4834       1   21        1                     0  33.001813 -117.263628   

      Number of Referrals  Tenure in Months  Offer  \
0                       2                 9      0   
1                       0                 9      0   
2                       0                 4      5   
3                       1                13      4   
4                       3                 3      0   
...                   ...               ...    ...   
4830                    0                 1      5   
4831                    0                13      4   
4832                    1                22      4   
4833                    0                 2      5   
4834                    5                67      1   

      Avg Monthly Long Distance Charges  ...  Unlimited Data  Contract  \
0                                 42.39  ...               1         1   
1                                 10.69  ...               0         0   
2                                 33.65  ...               1         0   
3                                 27.82  ...               1         0   
4                                  7.38  ...               1         0   
...                                 ...  ...             ...       ...   
4830                              42.09  ...               1         0   
4831                              46.68  ...               1         1   
4832                              16.20  ...               1         0   
4833                              18.62  ...               1         0   
4834                               2.12  ...               1         2   

      Paperless Billing  Payment Method  Monthly Charge  Total Charges  \
0                     1               1           65.60         593.30   
1                     0               1           -4.00         542.40   
2                     1               0           73.90         280.85   
3                     1               0           98.00        1237.85   
4                     1               1           83.90         267.40   
...                 ...             ...             ...            ...   
4830                  1               1           70.15          70.15   
4831                  0               1           55.15         742.90   
4832                  1               0           85.10        1873.70   
4833                  1               1           50.30          92.75   
4834                  0               1           67.85        4627.65   

      Total Refunds  Total Extra Data Charges  Total Long Distance Charges  \
0              0.00                         0                       381.51   
1             38.33                        10                        96.21   
2              0.00                         0                       134.60   
3              0.00                         0                       361.66   
4              0.00                         0                        22.14   
...             ...                       ...                          ...   
4830           0.00                         0                        42.09   
4831           0.00                         0         

In [6]:
y

0       1
1       1
2       0
3       0
4       0
       ..
4830    0
4831    1
4832    0
4833    1
4834    1
Name: Customer Status, Length: 4835, dtype: int64

$Step-4$ Train test Split

In [7]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=1234)
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(3868, 30)
(967, 30)
(3868,)
(967,)


$Step-5$ Read the base model

In [8]:
from sklearn.model_selection import GridSearchCV,cross_val_score
from sklearn.tree import DecisionTreeClassifier
grid_tree=DecisionTreeClassifier()#Base model
grid_tree

DecisionTreeClassifier()

$Step-6$ Create a Parameter file

In [9]:
grid_tree.get_params()

{'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'random_state': None,
 'splitter': 'best'}

In [10]:
param_grid = {
             'criterion':['gini','entropy'],
             'max_depth':[3,4,5,6,7,8],
             'min_samples_split':[2,3,4],
             'min_samples_leaf':[1,2,3,4],
             'random_state':[42,1234]
             }
#2*6*3*4*2=288

$Step-7$ Aplly Grid Search CV

In [11]:
import time
start=time.time()
grid_search=GridSearchCV(grid_tree,param_grid,scoring='accuracy',cv=5,verbose=True)

end=time.time()
print("Total time taken is:",(end-start))

Total time taken is: 0.0009987354278564453


In [12]:
grid_search

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [3, 4, 5, 6, 7, 8],
                         'min_samples_leaf': [1, 2, 3, 4],
                         'min_samples_split': [2, 3, 4],
                         'random_state': [42, 1234]},
             scoring='accuracy', verbose=True)

$Step-8$  Fit the data

In [13]:
import time
start=time.time()
grid_search.fit(X_train,y_train)
end=time.time()
print("Total time taken is:",(end-start))

Fitting 5 folds for each of 288 candidates, totalling 1440 fits
Total time taken is: 200.361581325531


$Step-9$ Get the Best Parameters

In [17]:
grid_search.best_estimator_

DecisionTreeClassifier(criterion='entropy', max_depth=7, min_samples_leaf=4,
                       random_state=42)

In [15]:
grid_search.best_score_

0.7916202854077037

In [16]:
grid_search.best_params_

{'criterion': 'entropy',
 'max_depth': 7,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 42}

Cross validation Score

In [18]:
accuarcy_list=cross_val_score(grid_search.best_estimator_,X_train,y_train,scoring='accuracy')
accuarcy_list

array([0.80620155, 0.80878553, 0.77260982, 0.78654592, 0.7839586 ])

In [19]:
accuarcy_list.mean()

0.7916202854077037

- Identify the Predictions and metrics by using best parameters$

out of 1440 fittings, one best fit model is outcome
- best fit means best Parameters
- for me best prameter combination is
{'criterion': 'entropy',
 'max_depth': 7,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 42}
- entropy
- depth=7
- min_sample-leaf=4
- min_samples_split= 2
- random_state= 42

In [20]:
from sklearn.tree import DecisionTreeClassifier
dtree=DecisionTreeClassifier(criterion='entropy',
                             max_depth= 7,
                             min_samples_leaf= 4,
                             min_samples_split= 2,
                             random_state= 42)
dtree.fit(X_train,y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=7, min_samples_leaf=4,
                       random_state=42)

In [21]:
y_pred_dt=dtree.predict(X_test)

In [22]:
y_pred_dt

array([1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1,
       1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1,
       0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1,
       0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1,
       0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1,
       0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1,
       1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0,
       1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0,

In [23]:
from sklearn.metrics import accuracy_score,precision_score,f1_score,recall_score
acc_dt=round(accuracy_score(y_test,y_pred_dt)*100,2)
precision_dt=round(precision_score(y_test,y_pred_dt),2)
recall_dt=round(recall_score(y_test,y_pred_dt),2)
f1_dt=round(f1_score(y_test,y_pred_dt),2)
print("Accuracy",acc_dt)
print("precision Score",precision_dt)
print("Recall score",recall_dt)
print("F1 Score",f1_dt)

Accuracy 79.73
precision Score 0.88
Recall score 0.82
F1 Score 0.85


In [24]:
dtree.feature_importances_

array([0.        , 0.08378202, 0.02902043, 0.04780145, 0.01960247,
       0.0110213 , 0.17454667, 0.04174015, 0.00646908, 0.01554073,
       0.        , 0.02104071, 0.01372138, 0.00412327, 0.        ,
       0.        , 0.00635753, 0.00577444, 0.01070719, 0.00297719,
       0.0061536 , 0.38810381, 0.        , 0.        , 0.04709674,
       0.0297168 , 0.        , 0.        , 0.00971828, 0.02498478])

In [28]:
imp_df=pd.DataFrame({"Feature Name":X_train.columns,"Importance":dtree.feature_importances_})
imp_df

Feature Name  Importance
0                              Gender    0.000000
1                                 Age    0.083782
2                             Married    0.029020
3                Number of Dependents    0.047801
4                            Latitude    0.019602
5                           Longitude    0.011021
6                 Number of Referrals    0.174547
7                    Tenure in Months    0.041740
8                               Offer    0.006469
9   Avg Monthly Long Distance Charges    0.015541
10                     Multiple Lines    0.000000
11                      Internet Type    0.021041
12            Avg Monthly GB Download    0.013721
13                    Online Security    0.004123
14                      Online Backup    0.000000
15             Device Protection Plan    0.000000
16               Premium Tech Support    0.006358
17                       Streaming TV    0.005774
18                   Streaming Movies    0.010707
19                    Streaming Music    0.002977
20                     Unlimited Data    0.006154
21                           Contract    0.388104
22                  Paperless Billing    0.000000
23                     Payment Method    0.000000
24                     Monthly Charge    0.047097
25                      Total Charges    0.029717
26                      Total Refunds    0.000000
27           Total Extra Data Charges    0.000000
28        Total Long Distance Charges    0.009718
29                      Total Revenue    0.024985

In [29]:
f1=imp_df.sort_values(by='importance',ascending=False)
f1

KeyError: 'importance'